[Sema] TypeWrappers: If type wrapper comes from `.swiftinterface` use synthesized decls

xedin opened the pull request:
Swift interfaces have both attribute and all publicly accessible
synthesized members, which means that the synthesis needs to be
taught to lookup existing synthesized declarations for `$storage`,
`$Storage`, and `init(storageWrapper:)` if request comes for
declaration that belongs to a Swift interface file.

Note that this PR also allows type wrappers if they appear in the
swiftinterface file because not all use situations could be detected
by `usesFeature*` methods i.e. existential value uses of type wrapped
protocol.

Resolves:  rdar://103425758 

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] TypeWrappers: Always allow attributes that come from swiftinterfaces

Since type wrapper attributes could be applied to protocols and
the protocol types could be used as existential values, it's not
always possible to wrap all uses of feature into `#if $TypeWrappers`
block so let's allow use of attributes if they come from a Swift
interface file.

xedin committed:
[Sema] TypeWrappers: If type wrapper comes from `.swiftinterface` use synthesized decls

Swift interfaces have both attribute and all publicly accessible
synthesized members, which means that the synthesis needs to be
taught to lookup existing synthesized declarations for `$storage`,
`$Storage`, and `init(storageWrapper:)` if request comes for
declaration that belongs to a Swift interface file.

xedin committed:
[Sema] TypeWrappers: Availability of `init(storageWrapper:)` should match that of `$Storage`

xedin committed:
[Tests] TypeWrappers/NFC: Add a tests to round-trip type wrappers in swift interfaces

xedin commented:
@swift-ci please test

xymus commented on lib/Sema/CodeSynthesis.cpp:
```diff
@@ -1657,14 +1657,29 @@ ConstructorDecl *SynthesizeTypeWrappedTypeStorageWrapperInitializer::evaluate(
+  // if available.
+  {
+    auto parentSF = wrappedType->getDeclContext()->getParentSourceFile();
+    if (parentSF && parentSF->Kind == SourceFileKind::Interface) {
```
I'm wondering if it's worth restricting this logic to swiftinterfaces. Could it be useful for the user to define custom implementations for the type wrapper's backing logic? Keeping a similar behavior between sources and swiftinterface would be nice too.

xedin replied:
This is something we intentionally don't want users to mess with.

xymus commented on test/Serialization/type_wrapper_in_swiftinterface.swift:
```diff
@@ -0,0 +1,95 @@
+  }
+}
+
+//@Wrapper
```
Forgotten comment?

xedin replied:
Oops, thank you!

xymus reviewed (approved):
This looks good!

xedin commented:
@swift-ci please test

xedin committed:
[Sema] TypeWrappers: Request type wrapper attribute type eagerly

The type is needed for serialization so it has to be requested
and chached in type wrapper info.

xedin commented:
@swift-ci please test

drodriguez commented on test/Serialization/type_wrapper_in_swiftinterface.swift:
```diff
@@ -0,0 +1,93 @@
+// RUN:   -emit-module-path %t/PublicModule.swiftmodule \
+// RUN:   -emit-module-interface-path %t/PublicModule.swiftinterface \
+// RUN:   -emit-private-module-interface-path %t/PublicModule.private.swiftinterface \
+// RUN:   -enable-experimental-feature TypeWrappers
```
Many of the tests that enable `TypeWrappers` seems to include `REQUIRES: asserts`. This test does not, and if one runs the test suite in a non-assert compiler, one gets the error "experimental feature 'TypeWrappers' cannot be enabled in a production compiler". Would it be correct adding `REQUIRES: asserts` also to this test?
